### ISSUE TITLE
Create a device that makes it really easy to live stream events

### ISSUE THREAD

**Issue by ajlennon** [DoESLiverpool/somebody-should#927]
### Overview

The recent MaintenanceFest event (#906) seemed a really successful learning experience for live streaming of events. I'm very keen on creating a live streaming infrastructure that's really easy for people to use.

This would then give us the options of

- streaming events at DoES Liverpool to the wider world (via Twitch, Facebook, Youtube Live, other)
- working with other groups to live stream their events at DoES Liverpool
- other use cases such as video conferencing, monitoring 3D print builds, building security etc.

I can see this helping to gain exposure for DoES in the local, national, and global maker communities.

I think it may also help to foster links with other Makerspaces to create a "Federation of Makerspaces" where we interact with each other, are aware of each others' capabilities and events, and so forth.

The needs for such an infrastructure are

- very easy to live stream an event
- low cost hardware for live streaming
- easy to remotely access and update firmware on the hardware device
- has to look reasonably professional to viewers (minimal "buffering", options for cuts between multiple video sources.
- accessible to all on the internet
- potential for monetisation?

### Device

- on the device side we can use a Raspberry Pi 3 which has Wifi and a GPU core for video encoding acceleration
- with this we can use either a PiCam or a webcam. I am currently using a USB Logitech webcam with built in audio.
- I am using Resin.io to manage firmware updates and remote access
- It is using GStreamer to stream RTMP streams to a configurable internet server 

I've made a start on this and have audio/video live streaming using the RTMP protocol

The GitHub repo is here: https://github.com/DynamicDevices/doescam

### Streaming infrastructure

- We're currently streaming to a Restreamer endpoint which allows us to stream on to Facebook, Twitch, YouTube and other endpoints
- [TBD] Need to look at using OBS Studio for real time editing

This also relates to potential for a local streaming server (#886) and supporting awesome AV (#685)

**Comment by ajlennon** [comment]
@MatthewCroughan 

>Do you think we can have that working for a raspberry pi zero? Does Resin make that easy? Not too familiar with it.

So Resin.io supports the Pi0 and Pi0W which is good news. That said even if it didn't we could potentially add support or use another firmware management infrastructure.

The problem is that the Pi0 is 1GHz single core whereas the Pi3 is quadcore 1.2GHz and more importantly the Pi0 does not have the OpenMax GPU which we use for accelerated video encoding.

We might not have the grunt we need with the Pi0 but there's nothing stopping us from investigating.

![image](https://user-images.githubusercontent.com/1537834/45943765-8f719d00-bfdf-11e8-8472-1b3b402572d4.png)

>Automatic gain control may also present a massive issue for audio mixing. If possible, when implementing audio, we should make it a priority to have those levels be as neutral as they can be. For USB microphones like the blue yeti, which would be a good thing to have for the community meetings, we can use alsamixer to control that over ssh, or control alsamixer via one of the gpio pins and a bash script.

I think we need a way to change a variety of configuration options in real time. We might do this with MQtt through a broker or some other way. Needs thought but is all very achievable I believe

**Comment by amcewen** [comment]
If you want an event to practice on, there'll be the [Making It!?](https://makingitconf.com/) conference here at DoES on the 12th October...

**Comment by ajlennon** [comment]
It's a date @amcewen !!! Also @jackie1050 was mentioning next weekend's "What Does Health Look Like" event

**Comment by MethodDan** [comment]
I may have agreed to help stream and test this stuff with @amcewen yesterday when we had a coffee hehe :)

Loving your work on this so far @ajlennon and @matthewcroughan

Looks you've stolen my thunder a bit there Alex and beat me to the punch. Actually you've savede a load of work but the PiStreamer legal team will still be in touch :D

**Comment by ajlennon** [comment]
Thanks for the kind words @MethodDan 😊 It's an "itch I've wanted to scratch" for some time as I think a really low cost, open, easy to use and maintain customisable Linux video streaming device would be really neat.

As with so many of my thoughts it took the hard work of @jackie1050 and @MatthewCroughan, coupled with chatting with you on Saturday to actually kick me into gear to so something concrete :) 

Also I've had chats on and off with @amcewen about all the great A/V stuff you and others in the community have been doing over the years so really keen to put something together as a team if there's interest!

**Comment by cheapjack** [comment]
Yes @ajlennon this is really good and great it's come out of that event...

I'm usually the least enthusiastic person for VR *but* just wondering if the pi system Alex made (leaving the pi0 discussion) could cope with streaming from a 360 video camera? Would not have to be live. Id really love to be able to observe how we make stuff in Google cardboard: and really would love to document bio material stuff that way. If we can do it I can justify buying a 360 camera for my research

If you watch our very own @jackie1050 making one of Laura's audio synths with people at the critical kits symposium in 2016 in cardboard mode it really gets across the texture of what you're making...

https://youtu.be/6PEeaHD4woU

**Comment by ajlennon** [comment]
@MethodDan 

>Would love to test this out and it seems mad that in 2018 there still isn't a decent open and cheap solution. We could publish our designs and code for others to build on.

Very much - the repo, basic as it is, is open and asap I will be documenting the process to get it all going. My hope is that other MakerSpaces could grab a RPi and a webcam, download the image and immediately start streaming their events...

[https://github.com/DynamicDevices/doescam](https://github.com/DynamicDevices/doescam)

**Comment by ajlennon** [comment]
@cheapjack 

>I'm usually the least enthusiastic person for VR but just wondering if the pi system Alex made (leaving the pi0 discussion) could cope with streaming from a 360 video camera?

You read my mind! I really want to see if we can do multiple cameras for 360 and stream straight up to YouTube which will handle 360... If the RPi can't handle it I am sure we could source hardware which could...

**Comment by MethodDan** [comment]
At this rate I could be presenting this at the Open Source Summit in Edinburgh come the end of October. No pressure (on me as well) hehe :)

**Comment by MethodDan** [comment]
Let's see where we get to with this, we have less than 4 weeks but if we have something that we want to present then I have no problem standing up in front of a crowd of a people to do so :)

**Comment by ajlennon** [comment]
I got one of @goatchurchprime 's Pi boards working with a PiCam today.

@MatthewCroughan if you can leave that Pi0w about for me I'll have a look at whether it has enough oopmh for what we need.

Ta.

**Comment by ajlennon** [comment]
I suppose for completeness we should have a RPi configuration that receives the streaming so people can plug in HDMI, power it on, and you've immediately got events streaming to your venue...?

**Comment by MatthewCroughan** [comment]
@ajlennon The one pi zero w I have is actually working as an ssh tunnel for me right now. I have two raspberry pi zero's, one has a broken radio, so no wifi, (but could probably solder a new one on, I tried but failed) and another non W one. Plenty of USB wifi adapters though, so will that suffice?

**Comment by ajlennon** [comment]
Yeah we can give a USB adaptor a go sure.

**Comment by ajlennon** [comment]
@MatthewCroughan how is OBS Studio architected? Is it a client-server model?

**Comment by ajlennon** [comment]
@cheapjack @MatthewCroughan @MethodDan this looks a great place to start with 360

https://www.raspberrypi.org/magpi/building-a-raspberry-pi-360-camera/

**Comment by MatthewCroughan** [comment]
@ajlennon it is client-server, although there are a lot of plugins that let you do other things that wouldn't be considered client-server.

**Comment by ajlennon** [comment]
ok so we could run the server side on a ProxMox box with the various streams going into it and then manage that from a laptop? Sounds perfect

**Comment by MatthewCroughan** [comment]
@ajlennon What CPU do you have inside of the VR pc? If it supports vt-d we can run two operating systems with almost no overhead, with GPU passthrough. A friend has done a lot of testing on the fx8350, the cpu in what was going to be the embedded-linux build server may not be quite powerful enough to reliably encode a 60fps 1080p stream, at least not with a window manager on top. We might be able to get away with it, in some circumstances, but it can drop frames. We need something with higher single core performance, and more physical cores to really trust that it won't drop frames.

**Comment by ajlennon** [comment]
I'm thinking I'd like to leave that set up for people to use for VR as is. If we can use your blade thingy that might be perfect?

**Comment by MatthewCroughan** [comment]
@ajlennon Sure, but you're going to have to convince people that it's worth putting in the server room. That would encode around 5-6 1080p streams without any issue

**Comment by ajlennon** [comment]
Ah well that's a different conversation :)

**Comment by MatthewCroughan** [comment]
I can bring the blade in today, if you'd like. Can store it under my desk for now.

**Comment by MatthewCroughan** [comment]
@ajlennon Will be in, in 30 mins, if you say yes to that. So choose wisely.

**Comment by ajlennon** [comment]
Sounds good!

**Comment by ajlennon** [comment]
Looking at Pi0W + PiCam support now.

Printing a case from here which looks a good starting point

https://www.thingiverse.com/thing:1709013

**Comment by amcewen** [comment]
I'd like us not to be bringing more and more stuff into the space in the assumption we can just hide it somewhere and ignore any concerns that others have about it.

/me looks at the UPS that's still cluttering up the meeting nook...

**Comment by ajlennon** [comment]
Good point - The UPS was going to go do a good home but I think somebody investigated and found the batteries don't work? That being the case I'll happy take it and skip it if nobody wants it?

**Comment by ajlennon** [comment]
OK @MatthewCroughan so I now have two devices running Gstreamer1.0/Resin and live streaming

- Pi3B with PiCam
- Pi0W with PiCam

I'm not convinced the Pi0W is reliable as it was dropping out but that might have been a power issue. Am going to leave it on a soak test.

So the next thing I think is to see if we can stream both channels into OBS Studio/ProxMox and see if we can do some editing and transformation

**Comment by ajlennon** [comment]
So we’ve run into our first hurdle. The streaming works well on the Pi3 but the same architecture falls over after up to 20 minutes on the Pi0w

Anybody fancy digging in and working out why? @MatthewCroughan?

**Comment by MethodDan** [comment]
I was dubious about whether the Pi0w would be powerful enough for this but maybe it can be hacked in some way. Think the Pi3 should be our main target, more expensive I know but we want something reliable. Just my 2 cents. Will be in DOES tomorrow afternoon to have a look at this with you guys :)

**Comment by ajlennon** [comment]
New enclosure...

![cam](https://user-images.githubusercontent.com/1537834/46091405-b5976880-c1aa-11e8-853a-5d1c5666e1a5.jpg)

**Comment by jackie1050** [comment]
Looks good! Could we consider a variant where the lens is covered when the camera isn't switched on?

**Comment by ajlennon** [comment]
Did some testing with @goatchurchprime's Pi0w and

- GStreamer pipeline with a test video source works fine
- Raspivid running by itself kernel oops's after a few minutes

Testing with @MatthewCroughan's Pi0w now and (famous last words) this seems to work ok.

So I'm going to put this down to a hardware fault for now

**Comment by ajlennon** [comment]
@MethodDan - yes I was very dubious as to whether a Pi0w would be up to the job but amazingly it does seem to be from my testing thus far.

@jackie1050 - yes something like that is needed isn't it. LED on/flashing when streaming?

**Comment by johnmckerrell** [comment]
Do you need to have the MPEG codec for all this or is it using H.264, and if you need the codec have you bought it for the Pi0w?

**Comment by ajlennon** [comment]
I am generating a RTMP which is being sent up to the cloud. This composes an FLV container with the h.264 stream muxed with an AAC audio stream.

The h,264 is described as "coming out of the pi cam" but I am not convinced that the camera is natively outputting h.264. Raspivid has options to set the MJPEG mode or the h.264 mode. 

I didn't think the Pi0 had a GPU so didn't think it could be doing encoding in hardware so maybe the stream is natively h.264 from the camera...

Ages ago I did buy a key for a Pi for some video coding/decoding but I haven't had to worry about buying any codec support for the Pi0w.

Still googling around to try to get to the bottom of the above. If you come up with anything please do let me know :)

**Comment by ajlennon** [comment]
OK so for background on codecs, I'm reading the following. So I guess for FLV + h.264 we're covered

> H264 is include in the base price of the Raspi's. The other codecs need to be paid for. This is because they are not free or opensource (actually, even people using the open source codecs on PC should be paying a licence fee). A company called MPEGLA charges for all commercial use of these codecs - the Raspberry Pi foundation HAS to pay them for any codec usage otherwise they will be sued.

**Comment by ajlennon** [comment]
There's a RED LED as in the picture here. Will have a look at how to make this visible outside the case

![red led](https://user-images.githubusercontent.com/1537834/46098168-a967d700-c1bb-11e8-84cb-26260ef2096b.jpg)

**Comment by johnmckerrell** [comment]
You say that Pi0 has no GPU so maybe it can't do hardware encoding/decoding key or not?

**Comment by ajlennon** [comment]
Ok more digging @johnmckerrell. I think that table I posted above and which I was going from is  wrong.

Looking here the Pi Zero hsa a GPU with suppor for 1080p30 H.264/MPEG-4 AVC high-profile decoder and encode

https://en.wikipedia.org/wiki/Raspberry_Pi

I'm 99.99% sure this is what is doing the encoding as I can't believe the camera sensor has h.264 encoding built in at the price. And I can't believe we are seeing the FPS we see with software encoding

**Comment by MatthewCroughan** [comment]
Created a systemd service for the compiled nginx on the vm. Config is located at `/usr/local/nginx/conf` 
[This article](https://obsproject.com/forum/resources/how-to-set-up-your-own-private-rtmp-server-using-nginx.50/) logs the process for setting up the rtmp with nginx.

**Comment by MethodDan** [comment]
Great work @MatthewCroughan nice one. Can you do me a favour and snapshot or backup that VM? Just so we can have it ready and don't need to rebuild later. We made serious progress yesterday between the 3 of us. I also just ordered another Pi0w and camera, should be here by early next week at latest. Go team!

**Comment by MethodDan** [comment]
Just a quick thought but I need to find a talk for LivLUG next week and this could be perfect. How would you guys feel about that? @MatthewCroughan you're usually at LivLUG anyway and I know @ajlennon is a Linux-y guy too. I'm happy to do most of the talking if you want or just sit back and let you guys do it. Maybe we could demo what we've been working on for the group. Video streaming with open hardware and Linux :)

**Comment by ajlennon** [comment]
Sounds a good plan to me

**Comment by ajlennon** [comment]
Having a lot of trouble here getting the RPi Zero to see connected USB devices. Not sure what's going wrong. That black cable of yours works for you doesn't it @MatthewCroughan

**Comment by ajlennon** [comment]
@MatthewCroughan @MethodDan 

Still having trouble in two areas with RPi Zero W (USB and OpenGL shaders)

For now I've done a basic write-up of how to get going with RPiv3w and a webcam

https://github.com/DynamicDevices/doescam/wiki/Setting-up:-Raspberry-Pi-3-W-with-USB-Webcam-(Local---not-Resin.io)

**Comment by MatthewCroughan** [comment]
@ajlennon, yeah, that black cable does work. As for OpenGL, enable full kms in `raspi-config` under advanced and gpu settings. Also worth nothing that the soc on the raspi zeros are identical to the original pi 1, so you can prototype on that because of the usb ports being on board.

**Comment by ajlennon** [comment]
Setting full kms helps local performance a lot - thanks for the tip. Can't seem to manage to get it streamed though. Grumble.

**Comment by MatthewCroughan** [comment]
If you stream with a really low encoding preset, but a high bandwidth to the rtmp server, it will transcode the footage and output it at specifically 30fps 4000kbps

**Comment by MatthewCroughan** [comment]
@ajlennon https://gstreamer.freedesktop.org/releases/gst-omx/1.0.0.html after a bit of research, it seems like the GPU mightn't be doing the encoding, and that the CPU might instead be being used because of a lack of a command line parameter. I recall seeing 70~ CPU usage last time you ran `top`. I wouldn't be at all surprised if it were using CPU to do the job, since the pi zero can run x11vnc quite well too, it's quite surprising.

**Comment by ajlennon** [comment]
Now with OpenGL fisheye effect

![img_2182](https://user-images.githubusercontent.com/1537834/46244902-c7148680-c3dd-11e8-9379-717baa16f462.jpg)

![img_2183](https://user-images.githubusercontent.com/1537834/46244904-ca0f7700-c3dd-11e8-9da5-88a07fb12dca.jpg)

**Comment by ajlennon** [comment]
@MatthewCroughan I don't think it is using the CPU, but if it was we would want to use the omxh264enc plugin for gstreamer.

**Comment by DoESsean** [comment]
Is this closeable? It seems like we're streaming pretty well now, but I don't know if you have everything you want yet.

**Comment by DoESsean** [comment]
Where are we with this @ajlennon?

**Comment by ajlennon** [comment]
We're still doing it but it's been proving a LOT harder than it appeared at first. 

It's bubbling along but will take time.

**Comment by ajlennon** [comment]
Closing this!
